spec.html

```diff
@@ -13799,6 +13892,74 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
+      <emu-clause id="sec-optional-chaining-evaluation">
+        <h1>Runtime Semantics: Evaluation</h1>
+        <emu-grammar>
+          OptionalExpression :
+            MemberExpression OptionalChain
+            CallExpression OptionalChain
+            OptionalExpression OptionalChain
+        </emu-grammar>
+        <emu-alg>
+          1. Let _baseExpression_ be the first child of this production (i.e., this |MemberExpression|, |CallExpression|, or |OptionalExpression|).
+          1. Let _baseReference_ be ? _baseExpression_.Evaluation().
+          1. Let _baseValue_ be ? GetValue(_baseReference_).
+          1. If _baseValue_ is *undefined* or *null*, then
+            1. Return *undefined*.
+          1. Let _optionalChain_ be this |OptionalChain|.
+          1. Return ? _optionalChain_.ChainEvaluation(_baseValue_, _baseReference_).
+        </emu-alg>
+      </emu-clause>
+      <emu-clause id="sec-optional-chaining-chain-evaluation">
+        <h1>Runtime Semantics: ChainEvaluation</h1>
+        <p>With parameters _baseValue_ and _baseReference_.</p>
+        <emu-grammar>OptionalChain : `?.` `[` Expression `]`</emu-grammar>
+        <emu-alg>
+          1. If the code matched by this production is strict mode code, let _strict_ be *true*, else let _strict_ be *false*.
+          1. Return ? EvaluateDynamicPropertyAccess(_baseValue_, |Expression|, _strict_).
+        </emu-alg>
+        <emu-grammar>OptionalChain : `?.` IdentifierName</emu-grammar>
+        <emu-alg>
+          1. If the code matched by this production is strict mode code, let _strict_ be *true*, else let _strict_ be *false*.
+          1. Return ? EvaluateStaticPropertyAccess(_baseValue_, |IdentifierName|, _strict_).
+        </emu-alg>
+        <emu-grammar>OptionalChain : `?.` Arguments</emu-grammar>
+        <emu-alg>
+          1. Let _thisChain_ be this production.
+          1. Let _tailCall_ be IsInTailPosition(_thisChain_).
+          1. Return ? EvaluateCall(_baseValue_, _baseReference_, |Arguments|, _tailCall_).
+        </emu-alg>
+        <emu-grammar>OptionalChain : OptionalChain `[` Expression `]`</emu-grammar>
+        <emu-alg>
+          1. Let _optionalChain_ be this |OptionalChain|.
+          1. Let _newReference_ be ? _optionalChain_.ChainEvaluation(_baseValue_, _baseReference_).
+          1. Let _newValue_ be ? GetValue(_newReference_).
+          1. If the code matched by this production is strict mode code, let _strict_ be *true*, else let _strict_ be *false*.
+          1. Return ? EvaluateDynamicPropertyAccess(_newValue_, |Expression|, _strict_).
+        </emu-alg>
+        <emu-grammar>OptionalChain : OptionalChain `.` IdentifierName</emu-grammar>
+        <emu-alg>
+          1. Let _optionalChain_ be this |OptionalChain|.
+          1. Let _newReference_ be ? _optionalChain_.ChainEvaluation(_baseValue_, _baseReference_).
+          1. Let _newValue_ be ? GetValue(_newReference_).
```

Comment:
Should there not be a

```
          1. If _newValue_ is *undefined* or *null*, then
            1. Return *undefined*.
```
after this line? (Ditto the ChainEvaluation sections for ``OptionalChain `[` Expression `]`` and ``OptionalChain : OptionalChain Arguments``.)

Reply:
No, if we have `a?.b.c`, then the chain is `?.b.c`—_whether_ we evaluate a chain is conditional but the evaluation of the chain is not.

Reply:
Oh, thank you, yes. If a `return` were necessary it would have occurred before this algorithm were reached.